Fix for issue #176, useref doesn't process all assets

asselin opened the pull request:
Depending on how much buffering was available in the gulp pipe, some assets would not be sent to the stage of the pipe.
Root cause is an error in API usage.  emit('end') was being called directly, instead of calling the transform stream's flush callback function.

Fixes #176

jonkemp commented:
Can you add a test that fails without this fix?

asselin commented:
I can take a crack, but it won't be today.

One way to test this is to pipe useref into a through2 stream that collects the filenames in the transform, and then in the flush function, compares the collected filenames to the list of expected filenames.  If the transform function inserts a delay before calling the callback, the files will queue up in useref, and not all of them will be passed to the through2 stream before it's flush is called.  To simplify, I'd also configure the through2 stream to have a small buffer (default IIRC is 16) so that you don't have to run useref with input that will create a ton of files.

jonkemp commented:
Thank you. And thanks for pointing this out. I'll take a look as well when I have some time.

asselin commented:
I fixed a few things in the PR update:
1. The tests were not consistent in how they compared filename paths, so the tests failed on Windows due to / vs \
2. The plugin didn't reinitialize some global variables when called again, so if a test failed, subsequent tests would timeout.
3. Added the test that shows how the assets wouldn't all get added to the stream if the buffer filled up.

Note: I also found a similar bug with the additionalStreams option.  I'll open a separate bug on that one.

jonkemp commented:
Thanks! Can you squash your commits?

jonkemp commented on lib/streamManager.js:
```diff
@@ -177,16 +177,20 @@ var exports = module.exports = (function () {
 
+            additionalFiles = [];
+            unprocessed = 0;
+            end = undefined;
```
Not sure why this is here.

asselin replied:
These are global variables used in gulp-useref that are not reinitialized if you invoke it a 2nd time.  See my comments from yesterday (#2)-- I ran into this problem when a test failed, it left useref in a partially completed state, and subsequent invocations all failed with a timeout because it wasn't reinitializing properly.

jonkemp replied:
How can I reproduce the problem that you are describing?

asselin replied:
You can repro by forcing a test to fail and seeing it's effect on subsequent tests.  For example, if you update 'should work with relative paths' to force a failure:

````javascript
// approx line 576 in test.js
        stream.on('data', function(newFile){
            should.exist(newFile.contents);
            if (a === 1) {
                'a'.should.equal('b');  // Force a failure
                path.normalize(newFile.path).should.equal(path.normalize('./test/fixtures/scripts/combined.js'));
            }
            ++a;
        });
````

You'll see that subsequent tests all time out

jonkemp commented on test/test.js:
```diff
@@ -177,9 +178,9 @@ describe('useref()', function() {
+                path.join(newFile.path).should.equal(path.join(__dirname, './fixtures/css/one.css'));
             } else if (a === 2) {
-                newFile.path.should.equal(path.join(__dirname, './fixtures/css/two.css'));
+                path.join(newFile.path).should.equal(path.join(__dirname, './fixtures/css/two.css'));
```
Shouldn't this be `path.normalize`?

asselin replied:
Either will work.  I assume you prefer path.normalize?

jonkemp replied:
Yes, I would prefer that since we are not joining anything here. :+1:

asselin commented:
Squashed and pushed.
